obspy/taup/seismic_phase.py

```diff
@@ -146,37 +152,68 @@ def parse_name(self, tMod):
                                 .format(self.name))
+
         # Set currWave to be the wave type for this leg, P or S
-        if currLeg in ("p", "K", "k") or currLeg[0] == "P":
+        if currLeg in ("p", "K", "k", "I") or currLeg[0] == "P":
             isPWave = True
+            isPWavePrev = isPWave
         elif currLeg in ("s", "J") or currLeg[0] == "S":
             isPWave = False
-        # First,  decide whether the ray is upgoing or downgoing from the
+            isPWavePrev = isPWave
+        else:
+            raise TauModelError('Unknown starting phase: ' + currLeg)
+
+        # First, decide whether the ray is upgoing or downgoing from the
         # source. If it is up going then the first branch number would be
         # model.sourceBranch-1 and downgoing would be model.sourceBranch.
+        upgoingRecBranch = tMod.findBranch(self.receiver_depth)
+        downgoingRecBranch = upgoingRecBranch - 1  # One branch shallower.
         if currLeg[0] in "sS":
             # Exclude S sources in fluids.
             sdep = tMod.source_depth
             if tMod.cmbDepth < sdep < tMod.iocbDepth:
                 self.maxRayParam, self.minRayParam = -1, -1
                 return
-        if currLeg[0] in "PS" or (self.expert and currLeg[0] in "KI"):
+
+        # Set self.maxRayParam to be a horizontal ray leaving the source and
+        # self.minRayParam to be a vertical (p=0) ray.
+        if currLeg[0] in "PS" or (self.expert and currLeg[0] in "KIJ"):
             # Downgoing from source.
             self.currBranch = tMod.sourceBranch
             # Treat initial downgoing as if it were an underside reflection.
-            self.endAction = self.REFLECTBOT
+            endAction = self.REFLECT_UNDERSIDE
+            try:
+                sLayerNum = tMod.sMod.layerNumberBelow(tMod.source_depth,
+                                                       isPWavePrev)
+                layer = tMod.sMod.getSlownessLayer(sLayerNum, isPWavePrev)
+                self.maxRayParam = layer['topP']
+            except SlownessModelError as e:
+                raise RuntimeError('Should not happen' + str(e))
+            self.maxRayParam = tMod.getTauBranch(tMod.sourceBranch,
+                                                 isPWave).maxRayParam
         elif currLeg in ("p", "s") or (self.expert and currLeg[0] == "k"):
             # Upgoing from source: treat initial downgoing as if it were a
             # topside reflection.
-            self.endAction = self.REFLECTTOP
+            endAction = self.REFLECT_TOPSIDE
+            try:
+                sLayerNum = tMod.sMod.layerNumberAbove(tMod.source_depth,
+                                                       isPWavePrev)
+                layer = tMod.sMod.getSlownessLayer(sLayerNum, isPWavePrev)
+                self.maxRayParam = layer['botP']
+            except SlownessModelError as e:
+                raise RuntimeError('Should not happen' + str(e))
```

Comment:
Maybe drop in a little "Please contact the developers."? (Same for all the other runtime errors)